app/controllers/users/two_factor_authentication_controller.rb

```diff
@@ -44,19 +49,56 @@ def validate_otp_delivery_preference_and_send_code
       end
     end
 
+    def update_otp_delivery_preference_if_needed
+      OtpDeliveryPreferenceUpdater.new(
+        user: current_user,
+        preference: delivery_params[:otp_delivery_preference],
+        context: otp_delivery_selection_form.context
+      ).call
+    end
+
     def handle_invalid_otp_delivery_preference(result)
       flash[:error] = result.errors[:phone].first
       preference = current_user.otp_delivery_preference
       redirect_to login_two_factor_url(otp_delivery_preference: preference)
     end
 
-    def invalid_phone_number(exception)
-      code = exception.code
-      analytics.track_event(
-        Analytics::TWILIO_PHONE_VALIDATION_FAILED, error: exception.message, code: code
+    def invalid_phone_number(exception, action:)
+      capture_analytics_for_exception(exception)
+
+      if action == 'show'
+        redirect_to_otp_verification_with_error
+      else
+        flash[:error] = error_message(exception.code)
+        redirect_back(fallback_location: account_url)
+      end
+    end
+
+    def redirect_to_otp_verification_with_error
+      flash[:error] = t('errors.messages.phone_unsupported')
+      redirect_to login_two_factor_url(
+        otp_delivery_preference: current_user.otp_delivery_preference, reauthn: reauthn?
       )
-      flash[:error] = error_message(code)
-      redirect_back(fallback_location: account_url)
+    end
+
+    # rubocop:disable Metrics/MethodLength
```

Comment:
This method is a bit long. What would you think about moving the top part into a `base_analytics_hash_for_exception` method or something like that?